sys/net/application_layer/gcoap/fileserver.c

```diff
@@ -264,13 +453,65 @@ static ssize_t gcoap_fileserver_directory_handler(coap_pkt_t *pdu, uint8_t *buf,
     return (uintptr_t)buf - (uintptr_t)pdu->hdr;
 }
 
+static ssize_t _put_directory(coap_pkt_t *pdu, uint8_t *buf, size_t len,
+                              struct requestdata *request,
+                              gcoap_fileserver_entry_t *resource)
+{
+    (void)resource;
+    int err;
+    vfs_DIR dir;
+    if (request->options.exists.if_match && request->options.if_match_len) {
+        return gcoap_fileserver_error_handler(pdu, buf, len, COAP_CODE_PRECONDITION_FAILED);
+    }
+    if ((err = vfs_opendir(&dir, request->namebuf)) == 0) {
+        vfs_closedir(&dir);
+        gcoap_resp_init(pdu, buf, len, COAP_CODE_CHANGED);
+    }
+    else {
+        if ((err = vfs_mkdir(request->namebuf, 0)) < 0) {
+            return gcoap_fileserver_error_handler(pdu, buf, len, err);
+        }
+        gcoap_resp_init(pdu, buf, len, COAP_CODE_CREATED);
+    }
+    return coap_opt_finish(pdu, COAP_OPT_FINISH_NONE);
+}
+
+static ssize_t _delete_directory(coap_pkt_t *pdu, uint8_t *buf, size_t len,
```

Comment:
Is there the possibility to signal recursive deletion?
Or is the CoAP fileserver API fundamentally low-level?

Reply:
Personally I am not interested in recursive deletion. Only empty directories can be deleted with `vfs_rmdir`,
If we want to implement recursive deletion later, we could define that a client must append a URI query `DELETE: coap://server:port/path/to/resource?recursive=1`. But this is just a first idea.

Reply:
I suggest that recursive is the default. The UNIX behavior can be easily emulated with ETags (just pick a special ETag for empty directories on the server side, and insist that the client send an `If-Match` if a directory would be removed recursively), whereas the other direction is cumbersome to emulate by the client.

Reply:
Do I understand this right?
If the client sends a `DELETE` on a directory, without an `If-Match`, the server should recursively delete the directory and all of its sub content.
If the client sends a `DELETE` on a directory, with an `If-Match` that contains the special Etag for empty directories, the server only deletes the directory if it is empty.